routing/index_graph_serializer.hpp

```diff
@@ -0,0 +1,381 @@
+    {
+      m_version = ReadPrimitiveFromSource<decltype(m_version)>(src);
+      if (m_version != kVersion)
+      {
+        MYTHROW(RoutingException,
+                ("Unknown index graph version ", m_version, ", current version ", kVersion));
+      }
+
+      m_numRoads = ReadPrimitiveFromSource<decltype(m_numRoads)>(src);
+      m_numJoints = ReadPrimitiveFromSource<decltype(m_numJoints)>(src);
+
+      auto const sectionsSize = ReadPrimitiveFromSource<uint32_t>(src);
+      m_sections.resize(sectionsSize);
+      for (Section & section : m_sections)
+        section.Deserialize(src);
+    }
+
+    uint32_t GetNumRoads() const { return m_numRoads; }
+    Joint::Id GetNumJoints() const { return m_numJoints; }
+    uint32_t GetNumSections() const { return m_sections.size(); }
+
+    Section const & GetSection(uint32_t index) const
+    {
+      CHECK_LESS(index, m_sections.size(), ());
+      return m_sections[index];
+    }
+
+    void AddSection(Section const & section) { m_sections.push_back(section); }
+
+  private:
+    uint8_t m_version = kVersion;
+    uint32_t m_numRoads = 0;
+    Joint::Id m_numJoints = 0;
+    vector<Section> m_sections;
+  };
+
+  class JointIdEncoder final
+  {
+  public:
+    template <class Sink>
+    void Write(Joint::Id originalId, BitWriter<Sink> & writer)
+    {
+      auto const & it = m_convertedIds.find(originalId);
+      if (it != m_convertedIds.end())
+      {
+        Joint::Id const convertedId = it->second;
+        if (convertedId >= m_count)
+        {
+          MYTHROW(RoutingException,
+                  ("Dublicated joint id should be less then count, convertedId =", convertedId,
+                   ", count =", m_count, ", originalId =", originalId));
+        }
+        writer.Write(1, 1);
+        Write32(writer, m_count - convertedId);
+        return;
+      }
+
+      m_convertedIds[originalId] = m_count;
+      writer.Write(0, 1);
+      ++m_count;
+    }
+
+    Joint::Id GetCount() const { return m_count; }
+
+  private:
+    Joint::Id m_count = 0;
+    unordered_map<Joint::Id, Joint::Id> m_convertedIds;
+  };
+
+  class JointIdDecoder final
+  {
+  public:
+    JointIdDecoder(Joint::Id begin) : m_count(begin) {}
+
+    template <class Source>
+    Joint::Id Read(BitReader<Source> & reader)
+    {
+      uint8_t const repeatBit = reader.Read(1);
+      if (repeatBit)
+      {
+        uint32_t const delta = Read32(reader);
+        if (delta > m_count)
+          MYTHROW(RoutingException, ("Joint id delta", delta, "> count =", m_count));
+
+        return m_count - delta;
+      }
+
+      return m_count++;
+    }
+
+    Joint::Id GetCount() const { return m_count; }
+
+  private:
+    Joint::Id m_count;
+  };
+
+  class JointIdConverter final
+  {
+  public:
+    JointIdConverter(Joint::Id numJoints) { m_ids.assign(numJoints, Joint::kInvalidId); }
+
+    Joint::Id Convert(Joint::Id jointIdInFile)
+    {
+      if (jointIdInFile >= m_ids.size())
+        MYTHROW(RoutingException,
```

Comment:
I guess there's no sense in exceptions during serialization, as serialization occurs on the server side. Feel free to directly CHECK here, as there's no way for generator to gracefully fix the error in the exception handler.